JBTM-3342 Propagate the parent context header (Long-Running-Action-Pa…

mmusgrov opened the pull request:
https://issues.redhat.com/browse/JBTM-3342

LRA
!MAIN !CORE !QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !AS_TESTS !TOMCAT !JACOCO

When propagating a nested context expose the parent context header on calls to methods carrying an LRA annotation.

xstefank commented on rts/lra/jaxrs/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java:
```diff
@@ -271,6 +271,28 @@ public void filter(ContainerRequestContext containerRequestContext) {
+                        // verify that parent corresponds to a valid URI
+                        new URI(parent);
+                        // the parent param was generated using URI.toASCIIString
+                        headers.putSingle(LRA_HTTP_PARENT_CONTEXT_HEADER, parent);
```
this will override user-provided header. Is it intended behavior?

mmusgrov replied:
Come on, read the spec, aren't we allowed to implement it. Does user behaviour override the spec now (ie does the users' opinion of what constitutes the transaction hierarchy trump actuality).

xstefank replied:
https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.java#L82

-- I think we should fail if the user propagates the wrong parent header, or at least log warning. This seems like it can hide potential issues with user code.

mmusgrov replied:
That part of the spec is referring to LRA_HTTP_CONTEXT_HEADER.
You can't randomly change the parent context since it forms a strict hierarchy.

xstefank commented on rts/lra/test/basic/src/test/java/io/narayana/lra/arquillian/NestedParticipantIT.java:
```diff
@@ -99,34 +106,71 @@ public void nestedParticipantAfterLRACallTest() {
+                    .path(NestedParticipant.ROOT_PATH)
+                    .path(NestedParticipant.PATH))
+                    .request()
+//                .header(LRA.LRA_HTTP_CONTEXT_HEADER, parentLRA)
```
Remove this please and add a comment explaining that the header is populated by the filter automatically?

mmusgrov replied:
Sure. I see that the fix has broken one of the tests so I will remove it at the same time I resolve that failure. I'll also add a test for deeper transaction hierarchies.

xstefank commented on rts/lra/test/basic/src/test/java/io/narayana/lra/arquillian/resource/NestedParticipant.java:
```diff
@@ -48,16 +47,33 @@
+    public Response runWithNestedContext(@HeaderParam(LRA.LRA_HTTP_CONTEXT_HEADER) URI lraId,
+                           @HeaderParam(LRA.LRA_HTTP_PARENT_CONTEXT_HEADER) URI parentId) {
+        if (parentId == null || lraId == null) {
+            return Response.serverError().build();
```
we should add a message that says which one was null. The error message in test says it is only parentId.

mmusgrov replied:
I don't think so, this is a test suite in which we already have tests that check the presence of LRA_HTTP_CONTEXT_HEADER.

xstefank replied:
true, this was more about this particular test only. It is unlikely it wouldn't be caught elsewhere.

mmusgrov replied:
This is a test for nested contexts so LRA_HTTP_CONTEXT_HEADER is not relevant.

mmusgrov commented:
@xstefank This is ready for another review now. I've fixed the broken test and added a test for deeper LRA hierarchies.
Please ignore the previous commits since I was using them as temporary backups and I forget that I had an outstanding PR. I'll delete the associated builds.

xstefank commented on rts/lra/service-base/src/main/java/io/narayana/lra/LRAConstants.java:
```diff
@@ -40,6 +40,8 @@
     public static final String CLIENT_ID_PARAM_NAME = "ClientID";
     public static final String TIMELIMIT_PARAM_NAME = "TimeLimit";
     public static final String PARENT_LRA_PARAM_NAME = "ParentLRA";
+    public static final String QUERY_PAIR_SEPARATOR = "&"; //
```
```suggestion
    public static final String QUERY_PAIR_SEPARATOR = "&";
```

xstefank commented on rts/lra/service-base/src/main/java/io/narayana/lra/Current.java:
```diff
@@ -64,6 +72,67 @@ public static Object getState(String key) {
+                        if (pa.length > 0) {
+                            if (decode) {
+                                return URLDecoder.decode(pa[0], "UTF-8");
+                            }
```
is it correct to return null if `decode` is false? In that case, this can be refactored to not contain the parameter.
